src/Compilers/Core/Portable/SourceGeneration/IncrementalContexts.cs

```diff
@@ -8,30 +8,40 @@
 using System.Threading;
 using Microsoft.CodeAnalysis.Diagnostics;
 using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.SourceGeneration;
 using Microsoft.CodeAnalysis.Text;
 
 namespace Microsoft.CodeAnalysis
 {
     /// <summary>
     /// Context passed to an incremental generator when <see cref="IIncrementalGenerator.Initialize(IncrementalGeneratorInitializationContext)"/> is called
     /// </summary>
-    public readonly struct IncrementalGeneratorInitializationContext
+    public readonly partial struct IncrementalGeneratorInitializationContext
     {
         private readonly ArrayBuilder<SyntaxInputNode> _syntaxInputBuilder;
         private readonly ArrayBuilder<IIncrementalGeneratorOutputNode> _outputNodes;
         private readonly string _sourceExtension;
 
-        internal IncrementalGeneratorInitializationContext(ArrayBuilder<SyntaxInputNode> syntaxInputBuilder, ArrayBuilder<IIncrementalGeneratorOutputNode> outputNodes, string sourceExtension)
+        internal readonly ISyntaxHelper SyntaxHelper;
+
+        internal IncrementalGeneratorInitializationContext(
+            ArrayBuilder<SyntaxInputNode> syntaxInputBuilder,
+            ArrayBuilder<IIncrementalGeneratorOutputNode> outputNodes,
+            ISyntaxHelper syntaxHelper,
+            string sourceExtension)
         {
             _syntaxInputBuilder = syntaxInputBuilder;
             _outputNodes = outputNodes;
+            SyntaxHelper = syntaxHelper;
             _sourceExtension = sourceExtension;
         }
 
-        public SyntaxValueProvider SyntaxProvider => new SyntaxValueProvider(_syntaxInputBuilder, RegisterOutput);
+        public SyntaxValueProvider SyntaxProvider => new(_syntaxInputBuilder, RegisterOutput, SyntaxHelper);
 
         public IncrementalValueProvider<Compilation> CompilationProvider => new IncrementalValueProvider<Compilation>(SharedInputNodes.Compilation.WithRegisterOutput(RegisterOutput).WithTrackingName(WellKnownGeneratorInputs.Compilation));
 
+        internal IncrementalValueProvider<CompilationOptions> CompilationOptionsProvider => new(SharedInputNodes.CompilationOptions.WithRegisterOutput(RegisterOutput).WithTrackingName(WellKnownGeneratorInputs.CompilationOptions));
```

Comment:
could make this public in the future as well.

Reply:
Turns out this creates a bit of a problem. In order to be part of a node the type has to be equatable and when you dig down into `CompilationOptions` it is not in the command line due to how our `MetadataReferenceResolver` are implemented. This is the stack trace I get when running this API via the command line compiler 

```
>	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.CommonCompiler.LoggingMetadataFileReferenceResolver.GetHashCode() Line 44	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.CommonCompiler.ExistingReferencesResolver.GetHashCode() Line 72	C#
 	Microsoft.CodeAnalysis.dll!Roslyn.Utilities.Hash.Combine<Microsoft.CodeAnalysis.MetadataReferenceResolver>(Microsoft.CodeAnalysis.MetadataReferenceResolver newKeyPart, int currentKey) Line 39	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.CompilationOptions.GetHashCodeHelper() Line 658	C#
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.CSharpCompilationOptions.GetHashCode() Line 753	C#
 	System.Private.CoreLib.dll!System.Collections.Generic.HashSet<Microsoft.CodeAnalysis.CompilationOptions>.AddIfNotPresent(Microsoft.CodeAnalysis.CompilationOptions value, out int location)	Unknown
 	System.Private.CoreLib.dll!System.Collections.Generic.HashSet<System.__Canon>.Add(System.__Canon item)	Unknown
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.InputNode<Microsoft.CodeAnalysis.CompilationOptions>.UpdateStateTable(Microsoft.CodeAnalysis.DriverStateTable.Builder graphState, Microsoft.CodeAnalysis.NodeStateTable<Microsoft.CodeAnalysis.CompilationOptions> previousTable, System.Threading.CancellationToken cancellationToken) Line 51	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.DriverStateTable.Builder.GetLatestStateTableForNode<Microsoft.CodeAnalysis.CompilationOptions>(Microsoft.CodeAnalysis.IIncrementalGeneratorNode<Microsoft.CodeAnalysis.CompilationOptions> source) Line 59	C#
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.TransformNode<Microsoft.CodeAnalysis.CompilationOptions, Microsoft.CodeAnalysis.SourceGeneration.GlobalAliases>.UpdateStateTable(Microsoft.CodeAnalysis.DriverStateTable.Builder builder, Microsoft.CodeAnalysis.NodeStateTable<Microsoft.CodeAnalysis.SourceGeneration.GlobalAliases> previousTable, System.Threading.CancellationToken cancellationToken) Line 43	C#
<lots of nodes ... />
```

Reply:
THe intent here is to just use ReferenceEquality for equality here.  I didn't realize that CompilatinoOptions implemented GetHashCode.  I'll just switch this to ref-comparison.

Reply:
Agree that is the right call. I did that in the setup I have locally and everything else worked fine.

Reply:
This isn't used anywhere other than inside `ForAttributeWithSimpleName`, so we can just use it directly in that method instead.